File: server/src/backtest.ts
/**
 * Historical Testing.
 *
 * Walks a historical OHLCV series bar-by-bar, re-using the same signal
 * detection and risk-management logic the live scanner uses, so backtest
 * results actually reflect the live strategy rather than a separate
 * approximation of it.
 */
import * as indicators from "./indicators";
import * as risk from "./risk";
import { classifyTrend } from "./regime";
import { detectSetup } from "./signals";
import { Candle } from "./marketData";

export interface TradeResult {
  entry_index: number;
  entry_price: number;
  exit_price: number;
  exit_reason: "take_profit" | "stop_loss" | "end_of_data";
  pnl_pct: number;
}

function round(v: number, digits: number): number {
  const factor = 10 ** digits;
  return Math.round(v * factor) / factor;
}

export class BacktestReport {
  trades: TradeResult[] = [];

  get winRate(): number {
    if (!this.trades.length) return 0.0;
    const wins = this.trades.filter((t) => t.pnl_pct > 0).length;
    return round((wins / this.trades.length) * 100, 2);
  }

  get avgWinPct(): number {
    const wins = this.trades.filter((t) => t.pnl_pct > 0).map((t) => t.pnl_pct);
    return wins.length ? round(wins.reduce((a, b) => a + b, 0) / wins.length, 2) : 0.0;
  }

  get avgLossPct(): number {
    const losses = this.trades.filter((t) => t.pnl_pct <= 0).map((t) => t.pnl_pct);
    return losses.length ? round(losses.reduce((a, b) => a + b, 0) / losses.length, 2) : 0.0;
  }

  get profitFactor(): number {
    const grossWin = this.trades.filter((t) => t.pnl_pct > 0).reduce((a, t) => a + t.pnl_pct, 0);
    const grossLoss = Math.abs(this.trades.filter((t) => t.pnl_pct <= 0).reduce((a, t) => a + t.pnl_pct, 0));
    if (grossLoss) return round(grossWin / grossLoss, 2);
    return grossWin ? Infinity : 0.0;
  }

  get maxDrawdownPct(): number {
    if (!this.trades.length) return 0.0;
    let equity = 0;
    let runningMax = -Infinity;
    let worstDrawdown = 0;
    for (const t of this.trades) {
      equity += t.pnl_pct;
      runningMax = Math.max(runningMax, equity);
      worstDrawdown = Math.min(worstDrawdown, equity - runningMax);
    }
    return round(worstDrawdown, 2);
  }

  get sharpeRatio(): number {
    if (this.trades.length < 2) return 0.0;
    const returns = this.trades.map((t) => t.pnl_pct);
    const mean = returns.reduce((a, b) => a + b, 0) / returns.length;
    const variance = returns.reduce((a, b) => a + (b - mean) ** 2, 0) / returns.length;
    const std = Math.sqrt(variance);
    if (std === 0) return 0.0;
    return round((mean / std) * Math.sqrt(returns.length), 2);
  }

  get maxConsecutiveLosses(): number {
    let streak = 0;
    let worst = 0;
    for (const t of this.trades) {
      if (t.pnl_pct <= 0) {
        streak += 1;
        worst = Math.max(worst, streak);
      } else {
        streak = 0;
      }
    }
    return worst;
  }

  summary() {
    return {
      num_trades: this.trades.length,
      win_rate_pct: this.winRate,
      avg_win_pct: this.avgWinPct,
      avg_loss_pct: this.avgLossPct,
      profit_factor: this.profitFactor,
      max_drawdown_pct: this.maxDrawdownPct,
      sharpe_ratio: this.sharpeRatio,
      max_consecutive_losses: this.maxConsecutiveLosses,
    };
  }
}

export function runBacktest(candles: Candle[], warmup = 60): BacktestReport {
  const report = new BacktestReport();
  let inTrade = false;
  let levels: risk.TradeLevels | null = null;
  let entryIndex = -1;

  for (let i = warmup; i < candles.length; i++) {
    const window = candles.slice(0, i + 1);

    if (inTrade && levels) {
      const bar = candles[i];
      if (bar.low <= levels.stop_loss) {
        const pnl = ((levels.stop_loss - levels.entry) / levels.entry) * 100;
        report.trades.push({
          entry_index: entryIndex,
          entry_price: levels.entry,
          exit_price: levels.stop_loss,
          exit_reason: "stop_loss",
          pnl_pct: round(pnl, 2),
        });
        inTrade = false;
      } else if (bar.high >= levels.take_profit_1) {
        const pnl = ((levels.take_profit_1 - levels.entry) / levels.entry) * 100;
        report.trades.push({
          entry_index: entryIndex,
          entry_price: levels.entry,
          exit_price: levels.take_profit_1,
          exit_reason: "take_profit",
          pnl_pct: round(pnl, 2),
        });
        inTrade = false;
      }
      continue;
    }

    let latest;
    try {
      latest = indicators.computeAll(window);
    } catch {
      continue;
    }

    const trend = classifyTrend(latest);
    const setup = detectSetup(latest, trend.label);
    if (setup === null) continue;

    levels = risk.computeLevels(setup.entry, setup.atr, setup.direction);
    if (!risk.meetsMinRR(levels)) continue;

    inTrade = true;
    entryIndex = i;
  }

  if (inTrade && levels) {
    const lastClose = candles[candles.length - 1].close;
    const pnl = ((lastClose - levels.entry) / levels.entry) * 100;
    report.trades.push({
      entry_index: entryIndex,
      entry_price: levels.entry,
      exit_price: lastClose,
      exit_reason: "end_of_data",
      pnl_pct: round(pnl, 2),
    });
  }

  return report;
}

File: server/app/backtest.py
"""Historical Testing.

Walks a historical OHLCV series bar-by-bar, re-using the same signal
detection and risk-management logic the live scanner uses, so backtest
results actually reflect the live strategy rather than a separate
approximation of it.
"""
from __future__ import annotations

from dataclasses import dataclass, field

import numpy as np
import pandas as pd

from . import indicators, risk
from .regime import classify_trend
from .signals import detect_setup


@dataclass
class TradeResult:
    entry_index: int
    entry_price: float
    exit_price: float
    exit_reason: str  # "take_profit" | "stop_loss" | "end_of_data"
    pnl_pct: float


@dataclass
class BacktestReport:
    trades: list[TradeResult] = field(default_factory=list)

    @property
    def win_rate(self) -> float:
        if not self.trades:
            return 0.0
        wins = sum(1 for t in self.trades if t.pnl_pct > 0)
        return round(wins / len(self.trades) * 100, 2)

    @property
    def avg_win_pct(self) -> float:
        wins = [t.pnl_pct for t in self.trades if t.pnl_pct > 0]
        return round(float(np.mean(wins)), 2) if wins else 0.0

    @property
    def avg_loss_pct(self) -> float:
        losses = [t.pnl_pct for t in self.trades if t.pnl_pct <= 0]
        return round(float(np.mean(losses)), 2) if losses else 0.0

    @property
    def profit_factor(self) -> float:
        gross_win = sum(t.pnl_pct for t in self.trades if t.pnl_pct > 0)
        gross_loss = abs(sum(t.pnl_pct for t in self.trades if t.pnl_pct <= 0))
        return round(gross_win / gross_loss, 2) if gross_loss else float("inf") if gross_win else 0.0

    @property
    def max_drawdown_pct(self) -> float:
        if not self.trades:
            return 0.0
        equity = np.cumsum([t.pnl_pct for t in self.trades])
        running_max = np.maximum.accumulate(equity)
        drawdown = equity - running_max
        return round(float(drawdown.min()), 2)

    @property
    def sharpe_ratio(self) -> float:
        if len(self.trades) < 2:
            return 0.0
        returns = np.array([t.pnl_pct for t in self.trades])
        if returns.std() == 0:
            return 0.0
        return round(float(returns.mean() / returns.std() * np.sqrt(len(returns))), 2)

    @property
    def max_consecutive_losses(self) -> int:
        streak = worst = 0
        for t in self.trades:
            if t.pnl_pct <= 0:
                streak += 1
                worst = max(worst, streak)
            else:
                streak = 0
        return worst

    def summary(self) -> dict:
        return {
            "num_trades": len(self.trades),
            "win_rate_pct": self.win_rate,
            "avg_win_pct": self.avg_win_pct,
            "avg_loss_pct": self.avg_loss_pct,
            "profit_factor": self.profit_factor,
            "max_drawdown_pct": self.max_drawdown_pct,
            "sharpe_ratio": self.sharpe_ratio,
            "max_consecutive_losses": self.max_consecutive_losses,
        }


def run_backtest(df: pd.DataFrame, warmup: int = 60) -> BacktestReport:
    """Re-runs the live setup-detection + risk logic bar-by-bar over
    historical OHLCV data. `warmup` bars are reserved for indicators to
    become valid before the first possible trade."""
    report = BacktestReport()
    in_trade = False
    levels = None
    entry_index = None

    for i in range(warmup, len(df)):
        window = df.iloc[: i + 1]

        if in_trade:
            bar = df.iloc[i]
            if bar["low"] <= levels.stop_loss:
                pnl = (levels.stop_loss - levels.entry) / levels.entry * 100
                report.trades.append(TradeResult(entry_index, levels.entry, levels.stop_loss, "stop_loss", round(pnl, 2)))
                in_trade = False
            elif bar["high"] >= levels.take_profit_1:
                pnl = (levels.take_profit_1 - levels.entry) / levels.entry * 100
                report.trades.append(TradeResult(entry_index, levels.entry, levels.take_profit_1, "take_profit", round(pnl, 2)))
                in_trade = False
            continue

        try:
            latest = indicators.compute_all(window)
        except ValueError:
            continue

        trend = classify_trend(latest)
        setup = detect_setup(latest, trend["label"])
        if setup is None:
            continue

        levels = risk.compute_levels(setup.entry, setup.atr, setup.direction)
        if not levels.meets_min_rr:
            continue

        in_trade = True
        entry_index = i

    if in_trade and levels is not None:
        last_close = df.iloc[-1]["close"]
        pnl = (last_close - levels.entry) / levels.entry * 100
        report.trades.append(TradeResult(entry_index, levels.entry, last_close, "end_of_data", round(pnl, 2)))

    return report
